File: julia-ing/LeetCode/300_LongestIncreasingSubsequence_DP.py
from bisect import bisect_left


# 1. common dynamic programming
class Solution1:
    def lengthOfLIS(self, nums):
        dp = [1] * len(nums)
        for i in range(len(nums)):
            for j in range(i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)


# 2. binary search
class Solution2:
    def lengthOfLIS(self, nums):
        tmp = [nums[0]]
        for x in nums[1:]:
            if tmp[-1] < x:  # x가 리스트 마지막 요소보다 크면
                tmp.append(x)
            else:
                idx = bisect_left(tmp, x)  # x를 삽입할 위치를 찾아 해당 위치의 값을 갈아 끼운다
                tmp[idx] = x
        return len(tmp)

File: eileenjang/LeetCode/300_LongestIncreasingSubsequence_DP.java
 // Dynamic Programming
class Solution {
    public int lengthOfLIS(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        int len = nums.length;
        int[] dp = new int[len];
        dp[0] = 1;
        int ans = 1;

        for (int i = 1; i < len; i++) {

            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
                else {
                    dp[i] = Math.max(dp[i], 1);
                }
            }

            ans = Math.max(ans, dp[i]);
        }

        return ans;
    }
}

// Binary Search
 public class Solution2 {
     public int lengthOfLIS(int[] nums) {
         int len = nums.length;
         int[] dp = new int[len];
         int ans = 0;

         for (int num : nums) {
             int idx = Arrays.binarySearch(dp, 0, ans, num);
             if (idx < 0) {
                 idx = -(idx + 1);
             }
             dp[idx] = num;
             if (idx == ans) {
                 ans++;
             }
         }

         return ans;
     }
 }
